File: packages/js/clinical/src/privacy.js
import {
  ALLOWED_ACT_KEYS,
  ALLOWED_CREATE_KEYS,
  MAX_USER_TEXT,
} from "./constants.js";
import { ClinicalError } from "./wire.js";

const EMAIL_RE = /[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,}/i;
const PHONE_RE = /(?:\+?\d{1,3}[\s.\-]*)?(?:\(?\d{3}\)?[\s.\-]*)\d{3}[\s.\-]*\d{4}/;
const SSN_RE = /\b\d{3}-\d{2}-\d{4}\b/;
const MRN_RE = /\b(mrn|niss|curp|dni|ssn)[:#\s]+\w+/i;
const ADDRESS_RE =
  /\b\d{1,5}\s+[A-Za-z0-9.'\-]+\s+(street|st|avenue|ave|road|rd|boulevard|blvd|lane|ln|drive|dr|calle|avenida)\b/i;
const NAME_RE =
  /\b(my name is|i am named|mi nombre es|i am)\s+[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(?:\s+[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)?\b/i;
const SYMPTOM_RE =
  /\b(i have\b.{0,60}\b(pain|fever|cough|symptoms?|chest|headache|nausea|chills)|i['’]m having\b.{0,40}\b(pain|fever|cough|symptoms?)|i am experiencing\b|me duele\b|tengo (dolor|fiebre|tos|sintomas|síntomas)\b|my (chest|head|stomach|throat|back) hurts\b|my symptoms\b|personal (symptom|history|information)\b)/i;
const BLOCKED_FIELDS = new Set([
  "symptoms",
  "personal_input",
  "phi",
  "patient_name",
  "full_name",
  "email",
  "phone",
  "address",
  "ssn",
  "dob",
  "mrn",
  "passport",
  "date_of_birth",
]);

function reject() {
  throw new ClinicalError("personal or free-text clinical input is not accepted", {
    code: "personal_input_rejected",
  });
}

export function looksPersonal(text) {
  const value = String(text || "");
  if (!value.trim()) return false;
  if (value.length > MAX_USER_TEXT) return true;
  return Boolean(
    EMAIL_RE.test(value)
      || PHONE_RE.test(value)
      || SSN_RE.test(value)
      || MRN_RE.test(value)
      || ADDRESS_RE.test(value)
      || NAME_RE.test(value)
      || SYMPTOM_RE.test(value),
  );
}

function walk(value) {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    for (const [key, item] of Object.entries(value)) {
      if (BLOCKED_FIELDS.has(String(key).toLowerCase()) && item) reject();
      walk(item);
    }
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) walk(item);
    return;
  }
  if (typeof value === "string" && looksPersonal(value)) reject();
}

export function rejectPersonalPayload(body, allowed) {
  if (!body || typeof body !== "object" || Array.isArray(body)) reject();
  const extra = Object.keys(body).filter((key) => !allowed.includes(key));
  if (extra.length) {
    if (extra.some((key) => BLOCKED_FIELDS.has(key.toLowerCase()))) reject();
    throw new ClinicalError("unsupported request field", { code: "invalid_request" });
  }
  for (const [key, value] of Object.entries(body)) {
    if (BLOCKED_FIELDS.has(String(key).toLowerCase()) && value) reject();
    if (typeof value === "string" && value.length > MAX_USER_TEXT) reject();
    walk(value);
  }
  return body;
}

export function rejectCreatePayload(body) {
  return rejectPersonalPayload(body, ALLOWED_CREATE_KEYS);
}

export function rejectActPayload(body) {
  return rejectPersonalPayload(body, ALLOWED_ACT_KEYS);
}

File: python/packages/handoffkit/handoffkit/clinical/privacy.py
"""Reject personal identifiers and free-text personal clinical input.

This is a predefined-case sandbox guard, not a perfect PHI detector.
Rejected text is never returned in errors, logs, or events.
"""

from __future__ import annotations

import re
from typing import Any

from handoffkit.clinical.constants import (
    ALLOWED_ACT_KEYS,
    ALLOWED_CREATE_KEYS,
    MAX_USER_TEXT,
)
from handoffkit.clinical.errors import ClinicalError

EMAIL_RE = re.compile(r"[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,}", re.I)
PHONE_RE = re.compile(r"(?:\+?\d{1,3}[\s.\-]*)?(?:\(?\d{3}\)?[\s.\-]*)\d{3}[\s.\-]*\d{4}")
SSN_RE = re.compile(r"\b\d{3}-\d{2}-\d{4}\b")
MRN_RE = re.compile(r"\b(mrn|niss|curp|dni|ssn)[:#\s]+\w+", re.I)
ADDRESS_RE = re.compile(
    r"\b\d{1,5}\s+[A-Za-z0-9.'\-]+\s+(street|st|avenue|ave|road|rd|"
    r"boulevard|blvd|lane|ln|drive|dr|calle|avenida)\b",
    re.I,
)
NAME_RE = re.compile(
    r"\b(my name is|i am named|mi nombre es|i am)\s+[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+"
    r"(?:\s+[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)?\b",
    re.I,
)
SYMPTOM_RE = re.compile(
    r"("
    r"\bi have\b.{0,60}\b(pain|fever|cough|symptoms?|chest|headache|nausea|chills)\b"
    r"|i['’]m having\b.{0,40}\b(pain|fever|cough|symptoms?)"
    r"|i am experiencing\b"
    r"|me duele\b"
    r"|tengo (dolor|fiebre|tos|sintomas|síntomas)\b"
    r"|my (chest|head|stomach|throat|back) hurts\b"
    r"|my symptoms\b"
    r"|personal (symptom|history|information)\b"
    r")",
    re.I,
)
BLOCKED_FIELDS = {
    "symptoms",
    "personal_input",
    "phi",
    "patient_name",
    "full_name",
    "email",
    "phone",
    "address",
    "ssn",
    "dob",
    "mrn",
    "passport",
    "date_of_birth",
}


def _reject() -> None:
    raise ClinicalError(
        "personal or free-text clinical input is not accepted",
        code="personal_input_rejected",
    )


def looks_personal(text: str) -> bool:
    value = str(text or "")
    if not value.strip():
        return False
    if len(value) > MAX_USER_TEXT:
        return True
    if EMAIL_RE.search(value):
        return True
    if PHONE_RE.search(value):
        return True
    if SSN_RE.search(value) or MRN_RE.search(value):
        return True
    if ADDRESS_RE.search(value):
        return True
    if NAME_RE.search(value):
        return True
    if SYMPTOM_RE.search(value):
        return True
    return False


def _walk(value: Any) -> None:
    if isinstance(value, dict):
        for key, item in value.items():
            if str(key).lower() in BLOCKED_FIELDS and item not in (None, False, "", 0):
                _reject()
            _walk(item)
        return
    if isinstance(value, list):
        for item in value:
            _walk(item)
        return
    if isinstance(value, str) and looks_personal(value):
        _reject()


def reject_personal_payload(body: dict[str, Any], *, allowed: frozenset[str]) -> dict[str, Any]:
    if not isinstance(body, dict):
        _reject()
    extra = set(body) - set(allowed)
    if extra:
        lowered = {str(key).lower() for key in extra}
        if lowered & BLOCKED_FIELDS:
            _reject()
        raise ClinicalError("unsupported request field", code="invalid_request")
    for key, value in body.items():
        if str(key).lower() in BLOCKED_FIELDS and value not in (None, False, "", 0):
            _reject()
        if isinstance(value, str) and len(value) > MAX_USER_TEXT:
            _reject()
        _walk(value)
    return body


def reject_create_payload(body: dict[str, Any]) -> dict[str, Any]:
    return reject_personal_payload(body, allowed=ALLOWED_CREATE_KEYS)


def reject_act_payload(body: dict[str, Any]) -> dict[str, Any]:
    return reject_personal_payload(body, allowed=ALLOWED_ACT_KEYS)
